### PR TITLE
Implement /msg command

### PR DESCRIPTION
**Pull request by sh0ckR6**
The `/msg [user] [message]` command opens a new channel with the specified user and immediately sends the given message to them.

Implements the usage of this command described by [the wiki](https://osu.ppy.sh/wiki/en/Chat_Console#/help). This command isn't implemented on stable to my knowledge, however.

### CODE REVIEW

File: osu.Game/Online/Chat/ChannelManager.cs [hunk 1]
```diff
@@ -256,8 +256,38 @@ public void PostCommand(string text, Channel target = null)
...
+                        OpenPrivateChannel(user);
+                        var message = new Message
+                        {
+                            Content = string.Join(" ", args[1..]),
+                            Sender = user,
+                            IsAction = false,
+                            ChannelId = CurrentChannel.Value.Id,
+                            Timestamp = DateTimeOffset.Now,
+                            DisplayContent = string.Join(" ", args[1..])
+                        };
```
Comment (jai-x) [inline osu.Game/Online/Chat/ChannelManager.cs:280]:
```suggestion
                    var args = content.Split(' ', 2);

                    var request = new GetUserRequest(args[0]);
                    request.Success += user =>
                    {
                        OpenPrivateChannel(user);
                        var message = new Message
                        {
                            Content = args[1],
                            Sender = user,
                            IsAction = false,
                            ChannelId = CurrentChannel.Value.Id,
                            Timestamp = DateTimeOffset.Now,
                            DisplayContent = args[1],
                        };
```
It's probably better not to oversplit and then rejoin.

Comment (sh0ckR6) [reply osu.Game/Online/Chat/ChannelManager.cs]: Makes sense, I've updated the PR with the changes.

**Comment by peppy** [comment]
suffers from same issue mentioned in #14637. also this is pretty inefficient and should be using `userLookupCache` or at least checking for an existing channel first.

**Review by rednir — CHANGES_REQUESTED** [review]

File: osu.Game/Online/Chat/ChannelManager.cs [hunk 2]
```diff
@@ -256,8 +256,38 @@ public void PostCommand(string text, Channel target = null)
...
                     break;
 
+                case "msg":
+                    if (string.IsNullOrWhiteSpace(content))
+                    {
+                        target.AddNewMessages(new ErrorMessage("Usage: /msg [user] [msg]"));
+                        break;
+                    }
+
+                    var args = content.Split(' ', 2);
```
[the remark this answers is not in the document]
Comment (sh0ckR6) [reply osu.Game/Online/Chat/ChannelManager.cs:266]: I feel like opening the channel without sending a message would be the right way to go about this instead of blocking the user's half-correct usage entirely.

**Comment by rednir** [comment]
This seems like almost the same functionality as #14637, just with an extra optional parameter. In my view it doesn't make sense to merge both.

One idea is to add aliases to this `/msg` command called `/chat` and `/query` (as these 3 commands have the same functionality stable). However one big drawback with this PR right now is not being able to specify a user with whitespaces in their username, such as "im a fancy lad".

**Comment by sh0ckR6** [comment]
> You could just add aliases to this /msg command called /chat and /query (these commands do the same thing in stable).

Would these aliases use their stable implementation suggested in #14637 or would they also have the extra optional parameter? I'm thinking if they're going to be used as aliases to `/msg`, they should also have the option to add an initial message. This is not what's implemented in stable, however, and is also not what's shown on [the wiki](https://osu.ppy.sh/wiki/en/Chat_Console#/help)

> However one big drawback with this PR right now is not being able to specify a user with whitespaces in their username, such as "im a fancy lad".

Yeah, I'm seeing this now. Without the optional parameter you could just pass all arguments into the name lookup and everything would work fine. With the message parameter added in, the line has to be drawn between the name and the message.

I'm trying to find a way to go about this without manually checking an increasing slice of arguments against the `UserLookupCache`/API, which would work but is also terribly inefficient.

**Comment by sh0ckR6** [comment]
Alright, I've done a bit of thinking and have come up with three options for continuing with this. If anyone has any other suggestions, please add them! I'm just trying to think out loud here and get some feedback.

### Option 1 - Stable Implementation
This option would just implement the `/msg` command as it is on stable. This would most likely be implemented as an alias on #14637, and this PR would be closed.

##### Example usage
`/msg [user]`

#### Pros
* Very simple to implement
* Users used to stable would know the command and expect it to work exactly the way it does on stable

#### Cons
* Would not implement the desired effect of starting a conversation with a user
* Wiki needs to be updated to reflect this implementation, as it currently shows that you can start a conversation with this command.

### Option 2 - Comma Delimitation
This option would have the user separate the target's name and message with a comma.

##### Example usage
`/msg [user], [message]`

#### Pros
* Simple to implement
* Supports both messages and usernames that contain spaces
* As far as I know, there are no other commands on stable that support multiple arguments, so trying out comma delimitation would not contradict stable users' previous experiences

#### Cons
* This may feel clunky to users as it may not be immediately obvious that they should use a comma
* Command systems implemented in other games (Minecraft, Counter-Strike: Global Offensive, etc.) do not use comma delimitation
  * Even if there has been no previous commands to contradict stable users' previous **experiences**, it may contradict their **expectations** from other games with similar systems

### Option 3 - Command System Upgrade
##### *This option is very different from the other options, and I would be more than willing to open a Discussion that talks about this in further detail if any project managers/maintainers are interested in taking osu!lazer down that path!
This option would be a heavy upgrade to the command system already in-place, making it somewhat like Discord's new "Slash Command" system.

Commands would essentially be C# objects that describe the command's shape. 
##### Example - New Command Creation
```csharp
var msgCommand = new Command {
  Name = "msg",
  Parameters = {
    new Parameter {
      Name = "user",
      Type = ParameterType.User,
      Required = true,
    },
    new Parameter {
      Name = "message",
      Type = ParameterType.String,
    },
  },
};
```

In-game there would be a UI similar to Discord's that would allow users to see the arguments required for the command. 
Autocomplete options could be included when initializing the `Command` object.
##### Example - `/msg [user] [name]` Command in Discord
/msg command

#### Pros
* Extremely user-friendly
* Built-in parameter verification
  * Type checking
  * Empty parameter (if marked as required)
* Easy-to-work-with command objects
  * Get parameters by name
  * Parameter separation
    * Multiple parameters that allow for whitespace will work
    * Because each parameter will be a variable, this gets built-in

#### General Notes
* Generalized solution
  * Can be applied to all current commands + all future commands
* Potentially allows for commands to be created by custom rulesets

#### Cons
* By far the longest solution to implement
* May require framework changes

### tl;dr
* It can be implemented the way it currently is on stable, becoming an alias of #14637 and closing this PR
* It can be implemented with comma delimitation)instead of space delimitation
* An entire command system upgrade could take place allowing for this to work as well as other features that make both command consumption by end-users and command development a better experience

Please remember that this is just me thinking out loud and looking for feedback, I'm not fully committed on any of these solutions and am very open for other suggestions. I am also more than willing to work on any of the above solutions as well as anything else that gets suggested!

**Comment by peppy** [comment]
This is all too much noise.

Implement as stable was. No huge discussion should be required.

**Comment by sh0ckR6** [comment]
Alright. I'll close this PR as it's now essentially a duplicate of #14637 and suggest `/msg` as an alias.
